pkg/controller/handlers/slackreceiver/slackreceiver.go

```diff
@@ -47,3 +74,168 @@ func CreateOAuthApp(req router.Request, _ router.Response) error {
 
 	return nil
 }
+
+func (s *Handler) SubscribeToSlackEvents(req router.Request, resp router.Response) error {
+	slackReceiver := req.Object.(*v1.SlackReceiver)
+
+	s.lock.RLock()
+	if _, ok := s.subscribed[slackReceiver.Name]; ok {
+		s.lock.RUnlock()
+		return nil
+	}
+	s.lock.RUnlock()
+
+	cred, err := s.gptScript.RevealCredential(req.Ctx, []string{system.OAuthAppPrefix + slackReceiver.Spec.ThreadName}, string(types.OAuthAppTypeSlack))
+	if err != nil {
+		return err
+	}
+
+	slackAppToken := cred.Env["APP_TOKEN"]
+
+	if slackAppToken == "" {
+		return nil
+	}
+
+	slackClient := slack.New("", slack.OptionAppLevelToken(slackAppToken))
+	client := socketmode.New(slackClient)
+
+	h := &eventHandler{
+		ctx:     req.Ctx,
+		storage: s.storage,
+	}
+
+	socketmodeHandler := socketmode.NewSocketmodeHandler(client)
+
+	socketmodeHandler.Handle(socketmode.EventTypeConnecting, middlewareConnecting)
+	socketmodeHandler.Handle(socketmode.EventTypeConnectionError, middlewareConnectionError)
+	socketmodeHandler.Handle(socketmode.EventTypeConnected, middlewareConnected)
+
+	// Handle a specific event from EventsAPI
+	socketmodeHandler.HandleEvents(slackevents.AppMention, h.middlewareAppMentionEvent)
+
+	ctx, cancel := context.WithCancel(req.Ctx)
+	s.lock.Lock()
+	s.subscribed[slackReceiver.Name] = cancel
+	s.lock.Unlock()
+
+	go func() {
+		if slackReceiver.Status.SocketConnectError != "" {
+			slackReceiver.Status.SocketConnectError = ""
+			if err := req.Client.Status().Update(req.Ctx, slackReceiver); err != nil {
+				log.Errorf("failed to update slack receiver status: %v", err)
+			}
+		}
+		err := socketmodeHandler.RunEventLoopContext(ctx)
+		if err != nil {
+			log.Errorf("error running event loop: %v", err)
+		}
+		s.lock.Lock()
+		if cancel, ok := s.subscribed[slackReceiver.Name]; ok {
+			cancel()
+			delete(s.subscribed, slackReceiver.Name)
+		}
+		s.lock.Unlock()
+		if err != nil {
+			err = fmt.Errorf("failed to run event loop: %w", err)
+			slackReceiver.Status.SocketConnectError = err.Error()
+			if err := req.Client.Status().Update(req.Ctx, slackReceiver); err != nil {
+				log.Errorf("failed to update slack receiver status: %v", err)
+			}
+		}
```

Comment:
Can we combine these two `if err != nil` checks? It doesn't look like `err` gets changed in between them.

Reply:
yeah, my bad. The first one can be moved down here